azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py

```diff
@@ -203,6 +217,39 @@ def _on_provider_publish_complete(self, mid):
             del self._event_callback_map[mid]
             callback()
 
+    def _on_provider_subscribe_complete(self, mid):
+        """
+        Callback that is called by the provider when a subscribe operation is complete.
+        """
+        logger.info("_on_provider_subscribe_complete")
+        callback = self._subscribe_callback
+        if callback:
+            self._subscribe_callback = None
+            callback()
+
+    def _on_provider_message_received_callback(self, topic, payload):
+        """
+        Callback that is called by the provider when a message is received.
+        """
+        logger.info(topic)
+        message_received = Message(payload)
+        # TODO : Discuss everything in bytes , need to be changed, specially the topic
+        topic_str = str(topic, "utf-8")
+        topic_parts = topic_str.split("/")
+
+        if "inputs" in topic_str and len(topic_parts) > 6:
```

Comment:
>6 [](start = 56, length = 1)

these feel like "magic numbers" to me.
constants might make this more readable
"TOPIC_POSITION_DEVICE_ID"
"TOPIC_POSITION_MODULE_ID"
"TOPIC_POSITION_ENCODED_PROPERTIES"
and so on

Reply:
you could also have helper functions that tell you what this topic is:
is_input_topic("<...>")
is_c2d_topic("<...>")
is_twin_topic("<...>")
etc

---
In reply to: [246900236](https://github.com/Azure/azure-iot-sdk-python-preview/pull/36#discussion_r246900236) [](ancestors = 246900236)